crates/backup-cli/src/adapters/p2p_client.rs

```diff
@@ -0,0 +1,142 @@
+use ed25519_dalek::{SigningKey, VerifyingKey};
+use mpc_contract::primitives::key_state::Keyset;
+use mpc_node::migration_service::web::client;
+
+use crate::{ports, types};
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("failed maket GET keyshares request: {0}")]
+    GetRequest(anyhow::Error),
+
+    #[error("failed maket PUT keyshares request: {0}")]
+    PutRequest(anyhow::Error),
+
+    #[error("failed maket PUT keyshares request: {0}")]
+    ServerConnection(anyhow::Error),
+}
+
+pub struct MpcP2PClient {
+    mpc_node_url: String,
+    mpc_node_p2p_key: VerifyingKey,
+    p2p_private_key: SigningKey,
+}
+
+impl MpcP2PClient {
+    pub fn new(
+        mpc_node_url: String,
+        mpc_node_p2p_key: VerifyingKey,
+        p2p_private_key: SigningKey,
+    ) -> Self {
+        Self {
+            mpc_node_url,
+            mpc_node_p2p_key,
+            p2p_private_key,
+        }
+    }
+}
+
+impl ports::P2PClient for MpcP2PClient {
+    type Error = Error;
+
+    async fn get_keyshares(&self, keyset: &Keyset) -> Result<types::KeyShares, Self::Error> {
+        let mut send_request = client::connect_to_web_server(
+            &self.p2p_private_key,
+            &self.mpc_node_url,
+            &self.mpc_node_p2p_key,
+        )
+        .await
+        .map_err(Error::ServerConnection)?;
```

Comment:
Nit: To be consistent with the previous review, we could technically improve our isolation of I/O here, but it would require refactoring `connect_to_web_server` and would be pretty complex, so I'd leave that as a follow-up exercise and not indulge in this exercise this time.

The cost is that our tests here actually will need to use I/O which means they can technically interfere or fail based on the host environment. However, that's not a big issue in practice and I would not expect any flakiness here.